packages/hardhat-core/src/internal/hardhat-network/provider/input.ts

```diff
@@ -77,8 +96,28 @@ export const logTopics = t.union([
 
 export type LogTopics = t.TypeOf<typeof logTopics>;
 
+export const blockTagObject = new t.Type<BN | Buffer>(
+  "BLOCKTAGOBJECT",
+  (u): u is BN | Buffer => BN.isBN(u) || Buffer.isBuffer(u),
+  (u, c) => {
+    if (isBlockTagObject(u)) {
+      if (u.blockHash != null) {
+        return t.success(toBuffer(u.blockHash));
+      }
+
+      if (u.blockNumber != null) {
+        return t.success(new BN(toBuffer(u.blockNumber)));
+      }
```

Comment:
Does the spec say what should happen if both `blockHash` and `blockNumber` are specified?

Reply:
Oh, I see that `isBlockTagObject` kind of checks that. I don't think we should do that though: an object with both values specified *is* a `BlockTagObject` (according to the interface). Throwing an error because that's invalid should be done elsewhere, maybe here. The result will be the same, it's just that the code will be a little clearer.

Reply:
Right. I am validating that currently by disallowing `blockHash` and `blockNumber` to be defined together. Only one is allowed as you observed. I didn't see an example of both `blockHash` and `blockNumber` in the same request payload in the EIP. Hence, my assumption is that it's not valid to set both in the same request payload.

I don't have a strong opinion on where the error should be thrown.

Reply:
Yes, I agree that an error should be thrown. It's just a minor comment about who is responsible for doing that: I think `isBlockTagObject` should return `true` if both are defined and then the decoder should fail if they are.

But this is _very_ minor and a personal preference, feel free to ignore it :sweat_smile:

Reply:
Gotcha. I mean you are right... In terms of "type" compliance with this current interface it would be fair to simply let it slide here because, yes, it's compliant with the TS interface if both are defined too. Maybe there is a third option though! We could create two different types (one for blockHash & one for blockNumber) and type them out explicitly. Would you prefer that or keep it simple?

Reply:
I think just handling it in the decoder is the balanced approach, the intention is clearer that way.